jax/experimental/pjit.py

```diff
@@ -0,0 +1,211 @@
+               for out, mesh_axes in safe_zip(out_nodes, out_mesh_axes)]
+
+  subc = subc.build(xops.Tuple(subc, out_nodes))
+  return xops.Call(c, subc, list(in_nodes))
+
+pjit_call_p = core.CallPrimitive("pjit_call")
+pjit_call_p.def_impl(_pjit_call_impl)
+xla.call_translations[pjit_call_p] = _pjit_translation_rule
+
+# None indicates replicated axis (partial replication)
+ArrayAxisPartitioning = Union[pxla.MeshAxisName, Tuple[pxla.MeshAxisName, ...], None]
+# None indicates fully replicated array value
+ArrayPartitioning = Optional[Tuple[ArrayAxisPartitioning, ...]]
+
+@lu.cache
+def _pjit_callable(
+    fun: lu.WrappedFun,
+    in_mesh_axes: Tuple[ArrayPartitioning, ...],
+    out_mesh_axes_thunk: Callable[[], Tuple[ArrayPartitioning, ...]],
+    resource_env,
+    name: str,
+    *in_avals):
+
+  in_axes = [get_array_mapping(axes) for axes in in_mesh_axes]
+  out_axes_thunk = lambda: [get_array_mapping(axes) for axes in out_mesh_axes_thunk()]
+  # TODO(skye): allow for using a submesh of physical_mesh
+  return pxla.mesh_tiled_callable(fun, name, None, resource_env.physical_mesh,
+                                  in_axes, out_axes_thunk, True, *in_avals,
+                                  fun_includes_mesh_axes=True)
+
+
+def with_sharding_constraint(x, mesh_axes):
+  # TODO(skye): this should arguably be part of axis_env
+  resource_env = maps.thread_resources.env
+  return sharding_constraint_p.bind(x, mesh_axes=mesh_axes,
+                                    resource_env=resource_env)
+
+def _sharding_constraint_impl(x, mesh_axes, resource_env):
+  # TODO(skye): can we also prevent this from being called in other
+  # non-pjit contexts? (e.g. pmap, control flow)
+  raise NotImplementedError(
+      "with_sharding_constraint() should only be called inside pjit()")
+
+def _sharding_constraint_translation_rule(c, x_node, mesh_axes, resource_env):
+  mesh = resource_env.physical_mesh
+  return xb.set_sharding_proto(c, x_node,
+                               get_sharding_proto(c, x_node, mesh_axes, mesh))
+
+sharding_constraint_p = core.Primitive("sharding_constraint")
+sharding_constraint_p.def_impl(_sharding_constraint_impl)
+sharding_constraint_p.def_abstract_eval(lambda x, **unused: x)
+ad.deflinear2(sharding_constraint_p,
+              lambda ct, _, mesh_axes, resource_env: (
+                  with_sharding_constraint(ct, mesh_axes),))
+xla.translations[sharding_constraint_p] = _sharding_constraint_translation_rule
+
+
+def get_array_mapping(mesh_axes: ArrayPartitioning) -> pxla.ArrayMapping:
+  if mesh_axes is None:
+    return OrderedDict()
+  return OrderedDict((axis, i)
+                     for i, axis_or_axes in enumerate(mesh_axes)
+                     for (axis, i) in _array_mapping_entries(axis_or_axes, i))
+
+def _array_mapping_entries(partitioning: ArrayAxisPartitioning, i: int):
+  if partitioning is None:
+    return
+  # TODO(skye): do axis names have to be strings?
```

Comment:
They don't. They can be anything and we only recurse when they're a `list` or a `tuple`.